src/isa.rs

```diff
@@ -0,0 +1,298 @@
+//!   ...
+//! end
+//! ```
+//!
+//! To execute `if_true_jump_to_end` , the interpreter needs to skip all instructions
+//! until it reaches the *matching* `end`. That's quite inefficient compared
+//! to a plain goto to the specific position.
+//!
+//! Because of this, the translation from the Wasm structured stack machine into a
+//! plain one is taking place.
+//!
+//! # Locals
+//!
+//! In a plain stack machine local variables and arguments live on the stack. Instead of
+//! accessing predifined locals slots in a plain stack machine locals are addressed relative
+//! to the current stack pointer. Because of this instead of taking an index of a local
+//! in {get,set,tee}_local operations, they take a relative depth as immediate. This works
+//! because at each instruction we always know the current stack height.
+//!
+//! Roughly, the stack layout looks like this
+//!
+//! | caller arguments |
+//! |  - arg 1         |
+//! |  - arg 2         |
+//! +------------------+
+//! | callee locals    |
+//! |  - var 1         |
+//! |  - var 2         |
+//! +------------------+
+//! | operands         |
+//! |  - op 1          |
+//! |  - op 2          |
+//! |                  |  <-- current stack pointer
+//! +------------------+
+//!
+//! # Differences from Wasm
+//!
+//! - There is no `nop` instruction.
+//! - All control flow strucutres are flattened to plain gotos.
+//! - Implicit returns via reaching function scope `End` are replaced with an explicit `return` instruction.
+//! - Locals live on the value stack now.
+//! - Load/store instructions doesn't take `align` parameter.
+//! - *.const store value in straight encoding.
+//! - Reserved immediates are ignored for `call_indirect`, `current_memory`, `grow_memory`.
+//!
+
+/// Should we keep a value before "discarding" a stack frame?
+///
+/// Note that this is a `enum` since Wasm doesn't support multiple return
+/// values at the moment.
+#[repr(u8)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum Keep {
+	None,
+	/// Pop one value from the yet-to-be-discarded stack frame to the
+	/// current stack frame.
+	Single,
+}
+
+/// Specifies how many values we should keep and how many we should drop.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct DropKeep {
+	pub drop: u32,
+	pub keep: Keep,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Target {
+	pub dst_pc: u32,
+	pub drop_keep: DropKeep,
+}
+
+#[allow(unused)] // TODO: Remove
+#[derive(Debug, Clone, PartialEq)]
+pub enum Instruction {
+	/// Push a local variable or an argument from the specified depth.
+	GetLocal(u32),
+
+	/// Pop a value and put it in at the specified depth.
+	SetLocal(u32),
+
+	/// Copy a value to the specified depth.
+	TeeLocal(u32),
+
+	/// Similar to the Wasm ones, but instead of a label depth
+	/// they specify direct PC.
+	Br(Target),
+	BrIfEqz(Target),
+	BrIfNez(Target),
+
+	/// Last one is the default.
+	///
+	/// Can be less than zero.
+	BrTable(Box<[Target]>),
```

Comment:
I dislike the pointer indirection here, it's not important to fix now but I wonder if we can do something about it. We could have `BrTable` have one field (the length) and then have a series of `Br` instructions afterwards that act as the list of targets.

Reply:
https://github.com/paritytech/wasmi/issues/100 !